Pilet setup on build

DanteDeRuwe opened the pull request:
This change will make sure the `Piral.Blazor.Tools` package transforms a regular Blazor app into a pilet.

#### Workflow:
- user creates blazor app `My.Blazor.Pilet`
- user adds `<PiralInstance>app-shell-name</PiralInstance>` as a property in his `My.Blazor.Pilet.csproj`
- user adds `Piral.Blazor.Tools` nuget package

#### Upon build of the Blazor project, `Piral.Blazor.Tools` will: 
- scaffold a pilet using the piral cli (`npx pilet new app-shell-name`)
- copy over a `blazor.codegen` (and pass on targetframework and configuration)
- modify the `index.tsx` to make use of this codegen
  - ```ts
    import { PiletApi } from 'app-shell-name';
    import { registerDependencies, registerBlazorPages, registerBlazorExtensions, setupPilet } from './blazor.codegen';
    
    export function setup(app: PiletApi) {
        registerDependencies(app);
        registerBlazorPages(app);
        registerBlazorExtensions(app);
        setupPilet(app);
    }
      ``` 
- modify the Blazor project's `wwwroot/index.html` for use with a development proxy
- copy over a kras script injector (and pass on targetframework and configuration) to act as a development proxy for IISExpress! 
  - This allows breakpoints to be triggered if a debugger is attached 🎉 
- install the `Piral.Blazor.Analyzer` tool locally for use in `blazor.codegen`

Piral-specific files are added outside the solution directory, in a `piral~` directory (that is by default not checked in thanks to the default .NET `.gitignore` file)

The user can create a `setup.tsx`  in the solution dir to register components on the pilet API, that are later integrated in the `index.tsx`.

#### Still to be done or considered:
- running `pilet debug` when Blazor application is debugged.
- changing the target framework, configuration or app shell name once the pilet is already scaffolded won't change the codegen. For this the `piral~` folder should be deleted and a rebuild has to be done

DanteDeRuwe committed:
Merge branch 'develop' into feature/tooling

DanteDeRuwe committed:
Add auto scaffolding and copying of files

DanteDeRuwe committed:
Overwrite index.html for debugging

for now on port 8080

DanteDeRuwe committed:
Overwrite index.tsx with custom one

DanteDeRuwe committed:
Change encoding to utf 8

Errors all over, this fixes it

DanteDeRuwe committed:
Define extra options in analyzer and package as tool

DanteDeRuwe committed:
Add blazor codegen

DanteDeRuwe committed:
Use Analyzer tool on build

DanteDeRuwe committed:
Install dependencies after scaffolding

To be sure

DanteDeRuwe committed:
Add extra metadata to analyzer

DanteDeRuwe committed:
Enable detection of optional setup.tsx file in Blazor project

FlorianRappl commented on src/Piral.Blazor.Tools/content/src/blazor.codegen:
```diff
@@ -200,6 +206,9 @@ module.exports = async function () {
     ${definePathCode}
     module.exports = {
+      setupPilet : (app) => {
+        ${setupFileExists ? `import("${setupFilePath}").then(e=> e.default(app))` : ''};
```
Why lazy load it? This is immediately required. Use `require('${setupFilePath}').default`.

```js
setupPilet: ${setupFileExists ? `require('${setupFilePath}').default` : '() => {}'
```

FlorianRappl commented on src/Piral.Blazor.Tools/content/src/blazor.codegen:
```diff
@@ -217,6 +226,9 @@ module.exports = async function () {
     return `
     ${definePathCode}
     module.exports = {
+      setupPilet : (app) => {
```
Hm why this duplicate?

Make it reusable with a function.

DanteDeRuwe committed:
Refactor blazor.codegen

DanteDeRuwe committed:
Fix typo

DanteDeRuwe committed:
Let MSbuild dynamically insert variables into the codegen

DanteDeRuwe committed:
Configure kras for debugging with IISExpress

Create a kras script injector that acts as a development proxy for IISExpress. Breakpoints can now be triggered in Visual Studio!

FlorianRappl commented on src/Piral.Blazor.Tools/content/src/mocks/debug.js:
```diff
@@ -0,0 +1,67 @@
+const piletManifest = require(bbjsonPath);
+
+const launchSettingsPath = resolve(blazorprojectfolder, 'Properties', lsjson);
+const iisUrl = require(launchSettingsPath)?.iisSettings?.iisExpress?.applicationUrl;
```
This wouldn't work in older Node versions (e.g., Node 10 and Node 12). Please write JS that works with Node 10 (or at least 12).

DanteDeRuwe replied:
That was my C# (and possibly TS) brain talking. Good catch. Changed it!

FlorianRappl reviewed (commented):
Looks great!

DanteDeRuwe committed:
Improve compatibility

https://github.com/FlorianRappl/Piral.Blazor/pull/14#discussion_r617792434
